### PR TITLE
feat(api-server): require github link for backend projects

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to: https://github.com/freeCodeCamp/freeCodeCamp/issues/53470

**MERGE/DEPLOY AFTER:** https://github.com/freeCodeCamp/freeCodeCamp/pull/53511

### CODE REVIEW

File: api-server/src/server/boot/challenge.js [hunk 1]
```diff
@@ -355,7 +355,10 @@ export function isValidChallengeCompletion(req, res, next) {
     log('challengeType', challengeType, isNumeric(challengeType));
     return res.status(403).json(isValidChallengeCompletionErrorMsg);
   }
-  if ('solution' in req.body && !isURL(solution)) {
+  const isUrlValid =
```
Comment (moT01) [inline api-server/src/server/boot/challenge.js:358]: Should we check if the GH link is a valid URL here too?

**Review by moT01 — COMMENTED** [review]
So for now, we are going to require both a solution URL and a source code link it looks like? I thought we wanted to make the solution link optional. Perhaps, in another PR or something.

It might be nice to add these changes to the new API here - but I'm fine with doing that later. Probably want to add this to the project board if we don't.

**Comment by ShaunSHamilton** [comment]
>  I thought we wanted to make the solution link optional. Perhaps, in another PR or something.

The solution link has to be submitted to run the tests. So, I thought it just made sense to keep it as is.

Also, it leaves it open for how we, in the future, want to do something with them.

**Comment by moT01** [comment]
This can be unblocked @ShaunSHamilton?

File: api-server/src/server/boot/challenge.js [hunk 2]
```diff
@@ -355,7 +355,18 @@ export function isValidChallengeCompletion(req, res, next) {
...
+  // - `githubLink` needs to exist and be valid URL
+  if (challengeType === challengeTypes.backEndProject) {
+    if (!solution || !githubLink) {
+      log('isObjectId', id, ObjectID.isValid(id));
+      return res.status(403).json(isValidChallengeCompletionErrorMsg);
+    } else if (!isURL(githubLink)) {
+      log('isObjectId', id, ObjectID.isValid(id));
+      return res.status(403).json(isValidChallengeCompletionErrorMsg);
+    }
+  } else if ('solution' in req.body && !isURL(solution)) {
```
Comment (moT01) [inline api-server/src/server/boot/challenge.js:369]: The comment above this says, `` `solution` needs to exist, but does not have to be valid URL`` - this appears to check that it is valid. Just want to make sure that's intended. I think it's fine.

Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:369]: > If `backEndProject`, `solution` needs to exist, but does not have to be valid. Hence:

```js
if (challengeType === challengeTypes.backEndProject)
```

**Review by moT01 — APPROVED** [review]
LGTM 🎉

File: api-server/src/server/boot/challenge.js [hunk 3]
```diff
@@ -355,7 +355,18 @@ export function isValidChallengeCompletion(req, res, next) {
...
+  // - `solution` needs to exist, but does not have to be valid URL
+  // - `githubLink` needs to exist and be valid URL
+  if (challengeType === challengeTypes.backEndProject) {
+    if (!solution || !githubLink) {
+      log('isObjectId', id, ObjectID.isValid(id));
+      return res.status(403).json(isValidChallengeCompletionErrorMsg);
+    } else if (!isURL(githubLink)) {
+      log('isObjectId', id, ObjectID.isValid(id));
+      return res.status(403).json(isValidChallengeCompletionErrorMsg);
+    }
```
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:368]:
```suggestion
    if (!solution || !isURL(githubLink + '')) {
      log('isObjectId', id, ObjectID.isValid(id));
      return res.status(403).json(isValidChallengeCompletionErrorMsg);
    }
```
This seemed a little verbose.

**Review by ojeytonwilliams — COMMENTED** [review]
I don't want to block because of a code-style issue, but also it's late for me, so I'm not the best person to approve.

File: api-server/src/server/boot/challenge.js (hunk 3 above)
Comment (raisedadead) [reply api-server/src/server/boot/challenge.js:368]: Why do we need the log?

Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:368]: No idea. We probably don't.

Comment (naomi-lgbt) [reply api-server/src/server/boot/challenge.js:368]: If `!isURL(githubLink + '')` is converting to a string, could we not `!isUrl(String(githubLink))`?

Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:368]: I have no idea about the log, but it was in all the others. So, I left it.

**Review by raisedadead — APPROVED** [review]

**Pull request merged**